linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameCompressor.scala

```diff
@@ -0,0 +1,227 @@
+    entries.getOrElseUpdate(fieldName, new FieldNameEntry(fieldName)).genResolver()
+
+  def genResolverFor(methodName: MethodName): () => String =
+    entries.getOrElseUpdate(methodName, new MethodNameEntry(methodName)).genResolver()
+
+  def genResolverFor(prop: ArrayClassProperty): () => String =
+    entries.getOrElseUpdate(prop, new ArrayClassPropEntry(prop)).genResolver()
+
+  /** Collects the property names to avoid for Scala instance members.
+   *
+   *  We collect the names of exported members in Scala classes. These live in
+   *  the same namespace as Scala methods and fields. Therefore, we must avoid
+   *  them when allocating names for that namespace.
+   */
+  private def collectPropertyNamesToAvoid(moduleSet: ModuleSet): Set[String] = {
+    import org.scalajs.ir.Trees._
+
+    val builder = Set.newBuilder[String]
+
+    builder ++= BasePropertyNamesToAvoid
+
+    for {
+      module <- moduleSet.modules
+      linkedClass <- module.classDefs
+      if linkedClass.kind.isClass
+      exportedMember <- linkedClass.exportedMembers
+    } {
+      (exportedMember: @unchecked) match {
+        case JSMethodDef(_, StringLiteral(name), _, _, _) =>
+          builder += name
+        case JSPropertyDef(_, StringLiteral(name), _, _) =>
+          builder += name
+      }
+    }
+
+    builder.result()
+  }
+}
+
+private[emitter] object NameCompressor {
+  /** Base set of names that should be avoided when allocating property names
+   *  in any namespace.
+   *
+   *  This set contains:
+   *
+   *  - the reserved JS identifiers (not technically invalid by spec, but JS
+   *    minifiers tend to avoid them anyway: `foo.if` is playing with fire),
+   *  - the `"then"` name, because it is used to identify `Thenable`s by
+   *    spec and therefore lives in the same namespace as the properties of
+   *    *all* objects,
+   */
+  private val BasePropertyNamesToAvoid: Set[String] =
+    NameGen.ReservedJSIdentifierNames + "then"
+
+  private def allocatePropertyNames(entries: EntryMap,
+      namesToAvoid: collection.Set[String]): Unit = {
+    val comparator: Comparator[PropertyNameEntry] =
+      Comparator.comparingInt[PropertyNameEntry](_.occurrences).reversed() // by decreasing order of occurrences
+        .thenComparing(Comparator.naturalOrder[PropertyNameEntry]()) // tie-break
+
+    val orderedEntries = entries.values.toArray
+    java.util.Arrays.sort(orderedEntries, comparator)
+
+    val generator = new NameGenerator(namesToAvoid)
+
+    for (entry <- orderedEntries)
+      entry.allocatedName = generator.nextString()
+  }
+
+  /** Keys of this map are `FieldName | MethodName | ArrayClassProperty`. */
+  private type EntryMap = mutable.AnyRefMap[AnyRef, PropertyNameEntry]
+
+  private sealed abstract class PropertyNameEntry extends Comparable[PropertyNameEntry] {
+    var occurrences: Int = 0
+    var allocatedName: String = null
+
+    protected def debugString: String
+
+    private object resolver extends (() => String) {
```

Comment:
Have you considered making the name resolver a trait in `DelayedIdent` (in the first commit already)?

It seems all usage sites need to use `object` syntax anyways to defined it. So we might as well define two abstract methods on it to avoid forgetting to implement it at the usage site (the `toString` of a lambda will be totally useless).

Reply:
It had actually occurred to me, but I did not went through.

I now introduced `DelayedIdent.Resolver` with `resolve()` and `debugString`.